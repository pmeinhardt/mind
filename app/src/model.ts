import type { Loro, LoroMap, LoroTree } from "loro-crdt";

export type Structure = {
  main: LoroTree<Node>;
  meta: LoroMap<Meta>;
};

export type Node = {
  expanded?: boolean;
  label?: string;
};

export type Meta = {
  name: string;
};

export function verify(doc: Loro): asserts doc is Loro<Structure> {
  // TODO: Verify document structure
  // const meta = doc.getMap("meta");
  // const main = doc.getTree("main");
}
