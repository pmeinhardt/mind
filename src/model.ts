import type { LoroMap, LoroTree } from "loro-crdt";
import { Loro } from "loro-crdt";

export type Structure = {
  main: LoroTree<Node>;
  meta: LoroMap<Meta>;
};

export type Node = {
  expanded?: boolean;
  label?: string;
};

export type Meta = {
  name: string;
};

export function create(name: string): Loro<Structure> {
  const doc = new Loro<Structure>();
  const meta = doc.getMap("meta");
  meta.set("name", name);
  return doc;
}

export function verify(doc: Loro): asserts doc is Loro<Structure> {
  // TODO: Verify document structure
  // const meta = doc.getMap("meta");
  // const main = doc.getTree("main");
}

export function read(
  file: File,
  { signal }: { signal?: AbortSignal } = {},
): Promise<Loro<Structure>> {
  return new Promise((resolve, reject) => {
    if (signal && signal.aborted) {
      reject(signal.reason);
      return;
    }

    const reader = new FileReader();

    signal?.addEventListener("abort", () => {
      reader.abort();
      reject(signal.reason);
    });

    reader.addEventListener("load", (event) => {
      const buffer = event.target?.result;

      if (typeof buffer === "undefined" || buffer === null) return;

      const bytes = new Uint8Array(buffer as ArrayBuffer);
      const doc = new Loro();

      try {
        doc.import(bytes);
        verify(doc);
        resolve(doc);
      } catch (error) {
        reject(error);
      }
    });

    reader.addEventListener("error", () => {
      reject(new Error("failed to read file"));
    });

    reader.readAsArrayBuffer(file);
  });
}
