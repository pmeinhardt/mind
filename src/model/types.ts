import type { LoroDoc, LoroMap, LoroTree } from "loro-crdt";

export type Doc = LoroDoc<Structure>;

export type Structure = {
  main: LoroTree<Node>;
  meta: LoroMap<Meta>;
};

export type Node = {
  expanded?: boolean;
  label: string;
};

export type Meta = {
  name: string;
};
