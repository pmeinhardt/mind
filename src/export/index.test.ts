import type { LoroTree, LoroTreeNode } from "loro-crdt";

import type { Node } from "../model";
import { Doc } from "../model";
import { markdown } from "./index";

const doc = (() => {
  const insert = (tree: LoroTree<Node> | LoroTreeNode<Node>, name: string) => {
    const node = tree.createNode();
    node.data.set("label", name);
    return node;
  };

  const doc = new Doc("Fruits");

  const apple = insert(doc.main, "Apple");
  insert(apple, "Boskoop");
  insert(apple, "Carola");
  insert(apple, "Elstar");

  const banana = insert(doc.main, "Banana");
  insert(banana, "Goldfinger");
  insert(banana, "Gros Michel");

  insert(doc.main, "Clementine");

  return doc;
})();

test("markdown", () => {
  expect(markdown(doc)).toMatchSnapshot();
});
