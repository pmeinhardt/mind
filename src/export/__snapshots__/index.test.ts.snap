// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdown 1`] = `
"# Fruits

- Apple
    - Boskoop
    - Carola
    - Elstar
- Banana
    - Goldfinger
    - Gros Michel
- Clementine"
`;
